### ISSUE TITLE
Need to add RowConvertible to non-GRDB aware class.

### ISSUE THREAD

**Issue by schveiguy** [groue/GRDB.swift#38]
I have a situation where I would like to separate out the mechanism used to load class instances using GRDB from the base type itself.

`RowConvertible` requires `init(_ row: Row)` function, but for non-final classes, this cannot be an extension, which means that I can't "tack on" compatibility with GRDB with an extension to something that has no knowledge of GRDB.

I think it would be good to be able to use a factory method which swift does not place extra requirements on, so you can apply GRDB compatibility completely using extensions.

Something like:

```Swift
protocol FactoryRowConvertible {
   static func loadFromRow(row: Row) -> Self
}
```

Apologize if I got this wrong syntax-wise, I'm somewhat new to Swift.

Then all the extensions added to `RowConvertible` could do the same with `FactoryRowConvertible`.

**Comment by groue** [comment]
Hello @schveiguy 

I completely understand your request. Actually, RowConvertible used to be a factory protocol, until I went back to the regular init in https://github.com/groue/GRDB.swift/commit/b85b9581270868331403425be8878128eb3eeafd.

The main reason for this change was that the implementation of init looked much nicer than the implementation of the factory method. Pure aestheticism :-) And I did not foresee your need.

I saw it quite clearly for *values* that can be stored in columns: Bool, Int, String, etc. In order to support NSData, NSData and other non-final classes, the [DatabaseValueConvertible](https://github.com/groue/GRDB.swift/blob/v0.55.0/GRDB/Core/DatabaseValueConvertible.swift#L22) is a factory protocol, exactly like the one you are suggesting.

For someone who is "somewhat new to Swift", you already have a very precise ideas of problems and their solutions :-)

I wish we could keep the simple RowConvertible. Maybe we could add a factory protocol on top of it, from which RowConvertible would inherit.

**Comment by schveiguy** [comment]
> And I did not foresee your need.

I did not either :) Actually, I was kind of surprised xcode was complaining about me trying to add GRDB support via extensions.

At this point, my wish is to try and separate out the usage of the objects while in my app from the storage/loading. But everything is under my control, so I can add the appropriate initializers if necessary. This was just an observation that it seemed like the requirements for having a protocol initializer were more strict than they need to be.

> For someone who is "somewhat new to Swift", you already have a very precise ideas of problems and their solutions :-)

I'm new to Swift as of October last year maybe? But not new to iOS programming, and certainly not new to programming in general ;) My comment is because I'm not sure of the syntax required, especially for protocols referring to types, and especially without Xcode to yell at me when I get it wrong...

I really like your library, I was going to use FMDB but I was realizing quite rapidly that the niceties of swift would be unavailable without severe work. In particular, having to wrap everything in NSNumber, etc. Figured, someone must have something better. This certainly fits the bill, great work!

> I wish we could keep the simple RowConvertible. Maybe we could add a factory protocol on top of it, from which RowConvertible would inherit.

If there's a way to have RowConvertible, and the FactoryRowConvertible type share implementation/API, that would be awesome. I think the possibility may be to add an extension to RowConvertible which adds the appropriate factory method (after all, you need to call an initializer from a factory method), and then use the factory method for implementing the extensions?

In other words, a RowConvertible can trivially be a FactoryRowConvertible by adding a simple factory extension. Perhaps with:

```Swift
extension RowConvertible : FactoryRowConvertible {
    func loadFromRow(row: Row) -> Self { return Self(row) }
}
```

Again, I'm sure I got the syntax wrong!

**Comment by groue** [comment]
Thank you for your compliments! It's always a pleasure to write code that is helpful.

> But everything is under my control, so I can add the appropriate initializers if necessary. 

Too late!

The [FactoryRowConvertible](https://github.com/groue/GRDB.swift/tree/FactoryRowConvertible) branch contains a FactoryRowConvertible that may suit your needs. Will you give it a try?

The API is otherwise unchanged, meaning that RowConvertible still exists and works as it used to do. Just, all methods and types that used to require RowConvertible now require FactoryRowConvertible. It's transparent for RowConvertible users, and FactoryRowConvertible is there for your use case.

```swift
public protocol FactoryRowConvertible {
    static func fromRow(row: Row) -> Self
    mutating func awakeFromFetch(row row: Row)
}

public protocol RowConvertible : FactoryRowConvertible {
    init(_ row: Row)
}

extension RowConvertible {
    public static func fromRow(row: Row) -> Self {
        return Self.init(row)
    }
}
```

Give a try to FactoryRowConvertible while I'm thinking about it, and tell me how it goes!

**Comment by schveiguy** [comment]
Nice!

Unfortunately, I can't get it to work, but not sure if I'm doing it wrong. My simple trial:
```Swift
import GRDB

class DataItem
{
    var a = 0
    var b = 0
}

extension DataItem : FactoryRowConvertible
{
    static func fromRow(row: Row) -> DataItem {
        let result = DataItem()
        
            result.a = row.value(named: "a") as Int
            result.b = row.value(named: "b") as Int
        return result
    }
}
```
Xcode reports:
```
Method 'fromRow' in non-final class 'DataItem' must return `Self` to conform to protocol 'FactoryRowConvertible'
```
I tried changing return type to `Self`, which comically allows the function, but then complains I can't convert `result` to it. I tried creating a var of type `Self`, Xcode complains I can only do that in a protocol, and it goes on :) Any ideas?

**Comment by schveiguy** [comment]
Bleh, looks like I have to mark the class final to implement the protocol. So we are stuck again on the same thing. Any ideas on how I can get it to work?

**Comment by schveiguy** [comment]
OK, got it to work:

```Swift
class DataItem
{
    var a = 0
    var b = 0
    required init() {}
}

extension DataItem : FactoryRowConvertible
{
    static func fromRow(row: Row) -> Self {
        let result = self.init()
        
            result.a = row.value(named: "a") as Int
            result.b = row.value(named: "b") as Int
        return result
    }
}
```

So I could potentially define DataItem somewhere (or hope that the class I want to instrument has a required initializer?) without it having to conform to GRDB protocols, and allow it to add the data.

It's still a little clunky, but has potential!

Thanks for the quick fix.

**Comment by groue** [comment]
Glad you found a "clunky" solution... But requiring an initializer is tough, isn't it?

When I look at the implementation of the DatabaseValueConvertible factory protocol for NSData, a non-final class, I see that it uses a [copy initializer](https://github.com/groue/GRDB.swift/blob/v0.55.0/GRDB/Core/Support/Foundation/NSData.swift#L19), and proceeds in two steps: 1. get a concrete NSData, 2. returns an instance of meta-type Self initialized from the concrete NSData. Thanks to Foundation writers, such copies don't copy much, actually.

But I could not make it work with your DataItem. Swift always finds a way to ask for required initialized. It looks like Objective-C classes have a blank check from the Swift compiler that our classes don't have (and I kind of understand it, given the mess around designated initializers in Objective-C).

Hmmm. The value of FactoryRowConvertible is quite impaired by the initializer requirement. I could still add it to GRDB - and avoid documenting it too much, but I couldn't help but thinking that Swift is still too young for such a pattern.

What do you think?

**Comment by schveiguy** [comment]
The changes are so minor, and took you what, like 5 minutes? I'm sure you can add it later if it becomes more useful.

For me, the required initializer would work, but I can see where this degrades the potential quite a bit (especially if you don't have control of the type you are looking to serialize). Almost seems like a mechanism external to the class makes sense, but that's likely a lot more invasive a change.

I'll probably just add the RowConvertible to my base class for now. This exercise was a bit academic, as I'm not sure if I'd ever use some other mechanism to load/store the data. I just like having that separation for some reason ;)

**Comment by groue** [comment]
> For me, the required initializer would work, but I can see where this degrades the potential quite a bit (especially if you don't have control of the type you are looking to serialize).

This is the whole point: if you control the type, you can add RowConvertible. If you don't control it, FactoryRowConvertible does not help (unless the type is a blessed Objective-C class, but I don't know the rules).

I agree with the academic side of the exercise, but this is how things go forward. Your request was quite sane, frankly!
